src/controllers/moderationactions/ModerationActionModel.cpp

```diff
@@ -1,28 +1,43 @@
 #include "controllers/moderationactions/ModerationActionModel.hpp"
 
 #include "controllers/moderationactions/ModerationAction.hpp"
+#include "messages/Image.hpp"
+#include "util/LoadPixmapLazy.hpp"
 #include "util/StandardItemHelper.hpp"
 
+#include <QIcon>
+#include <QPixmap>
+
 namespace chatterino {
 
 // commandmodel
 ModerationActionModel ::ModerationActionModel(QObject *parent)
-    : SignalVectorModel<ModerationAction>(1, parent)
+    : SignalVectorModel<ModerationAction>(2, parent)
 {
 }
 
 // turn a vector item into a model row
 ModerationAction ModerationActionModel::getItemFromRow(
     std::vector<QStandardItem *> &row, const ModerationAction &original)
 {
-    return ModerationAction(row[0]->data(Qt::DisplayRole).toString());
+    return ModerationAction(
+        row[Column::Command]->data(Qt::DisplayRole).toString(),
+        row[Column::Icon]->data(Qt::UserRole).toString());
 }
 
 // turns a row in the model into a vector item
 void ModerationActionModel::getRowFromItem(const ModerationAction &item,
                                            std::vector<QStandardItem *> &row)
 {
-    setStringItem(row[0], item.getAction());
+    setStringItem(row[Column::Command], item.getAction());
+    setFilePathItem(row[Column::Icon], item.iconPath());
+    if (!item.iconPath().isEmpty())
+    {
+        loadPixmapFromUrlLazy(
+            (*item.getImage())->url(), [row](const QPixmap &pixmap) {
+                row[Column::Icon]->setData(pixmap, Qt::DecorationRole);
+            });
```

Comment:
This should ensure the reference to `row` remains valid. Furthermore, `setData` should be called on the GUI thread.

> if iconPath is present, getImage will always be valid.

This should be noted in a comment.

Reply:
I don't think we can even check if `row` is valid. I can't make it a smart pointer without a large refactor. I could look up the row by some kind of unique temporary ID, but it seems like a hack.